Add runid

silanus23 opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #5036 |
| Primary OS tested on | Ubuntu|
| Robotic platform tested on |nav2_bringup|
| Does this PR contain AI generated software? | unit tests|
| Was this PR description generated by AI software? |Nope|

---

## Description of contribution in a few bullet points
I tried to implement the idea at #5036
I added run_id to the blackboard and it's symetrycal to action node template. So that everytime blackboard ticks nodes one can keep track of nodes init time. This is an insurance for synchronization in the background. Made this activated with a parameter so that it won't interfere with current implementations and break them too. 
<!--
-->

## Description of documentation updates required from your changes

Action nodes now have the parameter `is_global` for this feature.
<!--
-->

## Description of how this change was tested
I have tested this locally  with debug lines it was showing so integration tested too. Besides I wrote backward compatibility tests too.
<!--
-->

---

## Future work that may be required in bullet points
In future adding  current behavior trees `is_global = true` for every node 
As the final making this a default.
<!--
-->

#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

silanus23 committed:
backup

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
test fix

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
making last_run_id_ sentinel value adding more tests.

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
lint fixing and longer server time for test

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 commented:
@adivardi could you give this a review?  I just wondered if I comprehend the idea correctly.
Sorry to ping you directly. No worries if you are too busy.

adivardi commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -202,8 +204,28 @@ class BtActionNode : public BT::ActionNodeBase
+        }
+      } catch (const std::exception & e) {
+      // run_id not found on blackboard, use old behavior
+        if (!BT::isStatusActive(status())) {
```
I am not sure about this, that would be a weird surprise. Maybe just let it crash? It would be the same as if `server_timeout` is missing.

adivardi commented on nav2_behavior_tree/test/plugins/action/test_bt_action_node.cpp:
```diff
@@ -483,6 +489,180 @@ TEST_F(BTActionNodeTestFixture, test_server_cancel)
+
+  // Subsequent ticks with same RunID should continue without re-initialization
+  result = tree_->tickOnce();
+  EXPECT_EQ(result, BT::NodeStatus::RUNNING);
```
how come all tests have the same  `EXPECT_EQ(result, BT::NodeStatus::RUNNING);`  ?

adivardi reviewed (commented):
I left a few comments, but looks good!

We would need to add it to other nodes as well, all the ones that use `BT::isStatusActive()` to initialize (you can see https://github.com/ros-navigation/navigation2/pull/5035, though may new nodes were added in the meantime).

silanus23 commented:
Than you.

silanus23 commented:
@SteveMacenski  Sorry for late update. I couldn't give attention in here and other PRs. I am handling my hiring test case. Deadline is tomorrow. After that I beleive I can give others with this an answer in a week.

silanus23 committed:
adding uuid hashmap version

Signed-off-by: silanus23 <berkantali23@outlook.com>

SteveMacenski commented:
@adivardi can you review / close any resolved comments from your review?

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server.hpp:
```diff
@@ -21,6 +21,10 @@
 #include <string>
 #include <vector>
 
+#include <boost/uuid/uuid.hpp>
```
We need to link this / package.xml the dependency? Please only include just the boost libraries required (not all)

SteveMacenski commented on nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp:
```diff
@@ -77,6 +77,7 @@ class BtActionNode : public BT::ActionNodeBase
     if (getInput("server_name", remapped_action_name)) {
       action_name_ = remapped_action_name;
     }
+    getInput("is_global", is_global_);
```
Question: should this be applied to non-action nodes or only the action nodes?

SteveMacenski replied:
Also, I think this is also missing the other nodes that have the initialization procedure within their implementations themselves (not just the action server base node). I think that's a key part of the ticket https://github.com/ros-navigation/navigation2/issues/5036 but I can understand just starting in one spot to get it right before doing all the others!

silanus23 committed:
handling conflitcs and reviews of action nodes.

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
Deleting unnecessary complicated tests, adding necessary condition nodes and their tests

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 commented:
Currently I think I have added most of them. I will carry on batch by batch. 

Edit: When I start to write others I have realized that goal updated and battery nodes in current commit could need a fix please let me+ I have made `is_global` param node by node cause update conditions of these nodes and siutations can vary user to user.

silanus23 committed:
reverting is_battery_charging and is_battery_low to defaults, fixing goal_updated and single_trigger node updated.

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
Reverting unnecessary deletions.

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 commented:
@SteveMacenski  are these correct nodes? I have gone over the list and some of them might not need so I applied only to these nodes.

silanus23 commented:
@SteveMacenski  Don't want to ping you but can you give this a look please? How should I resolve conflict? Should I drop this runid on those nodes?

silanus23 commented:
Later looking back at conflicting PRs I have seen they are only commenting PRs, I thought that it was something that can deprecate ruin-id at first. So I have handled conflict. That's on me for being too presumptuous :/. 

I will try to send videos where changes log so review process will be easier. 

Edit: I later realized typos I will solve them

silanus23 committed:
take input on initialization

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
delete unnecessary boost includes

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
typo fix

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 commented:
https://drive.google.com/drive/folders/1_9RymXihTLZuSZ7DOTpwUuDvpsBrIR5c
in above time expired, goal updated and distance traveled nodes logging with runid change

https://drive.google.com/drive/folders/1_9RymXihTLZuSZ7DOTpwUuDvpsBrIR5c
in above single trigger node logging with runid change

I didn't add these nodes 
goal reached condition: not about runid it's about current place.

battery nodes + is_stopped: live conditions, independent from run

Transfrom available condition: This is latch based so runid would ruin this.

I guess these are only ones in that list. 

I will implement on remaining controller decorators too.

silanus23 committed:
fix single trigger node

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
add runid to ratecontroller

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
add runid to distancecontroller

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
change time expired and distance traveled into correct sementics

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
message standardization

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
add initialize function to needed nodes

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
add speed controller runid

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 commented:
OK I have now added controller nodes. I was not sure about the logic of time expired, distance traversed likes. I moved as said in the header. It  says it reinits with halts so I reinited them with new runid. I have added initialize to some of those node. 

I really wonder what you think about nodes I didn't add. They are diffirent than others even they are asynchronus. I am not sure about goal updated condition too. 

About should this be 1 yaml param or  unique xml port question: While implementing I have seen runid is effecting some of them in really diff aspects I would use xml port but that's a guess I am not sure if I can come up with a scenario.

Lastly I have tried every node I implemented with log system like I have sent videos above. They all work and robot moves.

silanus23 commented:
Hey before going offline I reviewed code again. I had seen some problems.
I am still not sure about goal nodes (goal updated condition, goal updated decorator, goal updater). I will wait for answer I can drop the goal updated's implementation. According to me no need since new goal is restarting every node again. 
I am not sure about no runid should throw std::runtime_error or BT::RunTimeError. 
I should put ``e.what()`` for throws. 
there are copy paste left over else blocks. I will delete them too :/

I will try to push these changes tomorrow.

silanus23 committed:
update to open throw messages and bt runtime error

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
commit fix

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
take back overfix

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 committed:
write inputs to node definitions

Signed-off-by: silanus23 <berkantali23@outlook.com>

silanus23 commented:
dropped goal updated cause a new run already comes with a new goal, so it already returns SUCCESS on its own runid wouldn't add anything, just suppress it. not 100% sure, can add it back if wanted. switched to bt error system as said in the prev msg I now take is global from bt input port so I wrote them to nav2_tree_nodes.xml.
